File: packages/voice/src/GeminiLiveVoiceSource.ts
import { insecureCredentialRefusal } from "./insecureCredential.js";
import { PcmAudioGraph } from "./PcmAudioGraph.js";
import { base64ToBytes, bytesToBase64, float32ToPcm16, parsePcmRate, pcm16ToFloat32 } from "./pcm.js";
import type { AgentState, VoiceMetrics, VoiceSource } from "@sinua/core";

/**
 * `VoiceSource` for Gemini Live over its raw WebSocket -- the PCM-over-
 * socket transport *shape* from docs/audio-pipeline.md's table (Amazon Nova
 * Sonic is the same shape over HTTP/2; ElevenLabs is the same shape with a
 * different protocol, see `ElevenLabsVoiceSource`). Unlike
 * `OpenAIRealtimeVoiceSource`, nothing hands us a playable track: the shared
 * `PcmAudioGraph` captures the mic and ships PCM16 @ 16 kHz, and decodes/
 * schedules the model's PCM16 @ 24 kHz chunks onto a playback timeline.
 *
 * Wire protocol, verified against the current reference
 * (ai.google.dev/api/live, camelCase JSON) and Google's own browser sample
 * (`google-gemini/live-api-web-console`), not recalled -- full citations in
 * from the vendor's public docs:
 *   - `wss://generativelanguage.googleapis.com/ws/google.ai.generativelanguage
 *     .v1beta.GenerativeService.BidiGenerateContent?key=…` (API key) or
 *     `…BidiGenerateContentConstrained?access_token=…` (ephemeral token);
 *   - first message `{"setup": {...}}` -> `{"setupComplete": {}}`;
 *   - mic: `{"realtimeInput": {"audio": {"data": <base64>, "mimeType":
 *     "audio/pcm;rate=16000"}}}`;
 *   - model: `serverContent.modelTurn.parts[].inlineData.{data, mimeType}`,
 *     plus `interrupted` / `turnComplete` / `generationComplete` /
 *     `waitingForInput` / `inputTranscription`; `goAway.timeLeft`;
 *     `sessionResumptionUpdate.{newHandle, resumable}`.
 *
 * ## The gate: `speaking` follows the playback timeline, not chunk receipt
 *
 * Chunks arrive well before they are audible. `PcmAudioGraph` schedules
 * every decoded chunk at a running cursor and hangs the analyser off the
 * playback gain node, so `AudioAnalysis` can only read what is audible and
 * `tick()` derives `speaking` from the graph's `playbackState()`.
 * `serverContent.interrupted` (user barged in) stops every scheduled source
 * and resets the cursor, as the Live guide instructs.
 *
 * ## State mapping (LiveKit `AgentState` vocabulary, ./types.ts)
 *
 * Gemini exposes **no** server-side user-VAD start/stop message (the
 * reference's `speechState` is deprecated with an undefined replacement;
 * LiveKit's and Pipecat's Gemini adapters both work around the same gap),
 * so, unlike OpenAI, "the user stopped talking" is never signalled. What is
 * signalled and used here: `setupComplete` -> `listening`; `interrupted`
 * -> `listening`; input transcription arriving (Gemini's own ASR heard the
 * user) -> `listening`; a chunk received but not yet audible, or the buffer
 * starved mid-generation -> `thinking`; audio audible -> `speaking`;
 * drained after `turnComplete`/`generationComplete`/`waitingForInput` ->
 * `listening`. `thinking` is therefore short by construction here.
 *
 * ## Session lifetime
 *
 * A connection lives ~10 minutes; the server sends `goAway` first. The
 * setup asks for `sessionResumption`, the latest `newHandle` is kept, and
 * on `goAway` or an unexpected close the adapter reopens the socket with
 * `sessionResumption: {handle}` (mic and playback graphs untouched, state
 * `initializing` meanwhile), the same loop LiveKit/Pipecat run.
 * `contextWindowCompression` is requested so the 15-minute audio-only
 * session cap doesn't apply either.
 *
 * ## Credentials -- read this before wiring it anywhere real
 *
 * - `auth_tokens/…` -> an ephemeral token minted by *your backend*
 *   (`POST /v1beta/auth_tokens`), sent as `access_token` on the Constrained
 *   endpoint. The production shape: short-lived, single-use by default,
 *   can lock the model/config server-side.
 * - anything else -> a raw API key on `?key=`. **DEV-ONLY DEMO WIRING**,
 *   same rule as `OpenAIRealtimeVoiceSource`: memory only, never persisted, never
 *   logged. It is in the socket URL only because a browser `WebSocket` can
 *   set no auth header and that is the API's documented key path -- one
 *   more reason a product must use the token path instead.
 */
export interface GeminiLiveVoiceSourceOptions {
  /** `auth_tokens/…` ephemeral token (production shape) or, dev-only, a raw API key. */
  credential: string;
  /**
   * Opt in to a raw, long-lived API key. Without it `connect()` refuses one
   * before touching the mic or the socket -- see `insecureCredential.ts`.
   * Local demos only (the Studio sets it); never in a shipped app.
   */
  allowInsecureApiKey?: boolean;
  /** Live model id; `gemini-3.8-live` is the current stable default per ai.google.dev/gemini-api/docs/models. */
  model?: string;
  /** Optional system instruction for the session. */
  instructions?: string;
}

const WS_BASE = "wss://generativelanguage.googleapis.com/ws/google.ai.generativelanguage.v1beta.GenerativeService.";
const METHOD_KEY = "BidiGenerateContent";
const METHOD_TOKEN = "BidiGenerateContentConstrained";
const INPUT_RATE = 16000;
const OUTPUT_RATE_FALLBACK = 24000;
const UPDATE_MS = 1000 / 30; // ~30fps metering, same as every other VoiceSource
const SETUP_TIMEOUT_MS = 15_000;
const RECONNECT_ATTEMPTS = 3;
const RECONNECT_DELAY_MS = 500;

interface ServerMessage {
  setupComplete?: unknown;
  serverContent?: {
    modelTurn?: { parts?: Array<{ inlineData?: { data?: string; mimeType?: string } }> };
    turnComplete?: boolean;
    interrupted?: boolean;
    generationComplete?: boolean;
    waitingForInput?: boolean;
    inputTranscription?: { text?: string };
    interimInputTranscription?: { text?: string };
  };
  sessionResumptionUpdate?: { newHandle?: string; resumable?: boolean };
  goAway?: { timeLeft?: string };
  error?: unknown;
}

export class GeminiLiveVoiceSource implements VoiceSource {
  private readonly credential: string;
  private readonly allowInsecureApiKey: boolean | undefined;
  private readonly model: string;
  private readonly instructions: string | undefined;

  private readonly graph = new PcmAudioGraph();
  /** `generationComplete`/`turnComplete`/`waitingForInput` seen since the last audio chunk. */
  private generationDone = true;

  private ws: WebSocket | null = null;
  private resumptionHandle: string | null = null;
  private wantConnected = false; // user intent -- false after disconnect()
  private connected = false;
  private streaming = false; // mic chunks are sent only after setupComplete
  private reconnecting = false;
  private intervalId: ReturnType<typeof setInterval> | null = null;
  private metricsCb: ((m: VoiceMetrics) => void) | null = null;
  private stateCb: ((s: AgentState) => void) | null = null;
  private interruptCb: (() => void) | null = null;
  private state: AgentState = "idle";

  constructor(opts: GeminiLiveVoiceSourceOptions) {
    this.credential = opts.credential.trim();
    this.allowInsecureApiKey = opts.allowInsecureApiKey;
    this.model = opts.model ?? "gemini-3.8-live";
    this.instructions = opts.instructions;
  }

  onMetrics(cb: (m: VoiceMetrics) => void): void {
    this.metricsCb = cb;
  }

  onStateChange(cb: (s: AgentState) => void): void {
    this.stateCb = cb;
  }

  onInterrupt(cb: () => void): void {
    this.interruptCb = cb;
  }

  async connect(): Promise<void> {
    if (!this.credential) throw new Error("GeminiLiveVoiceSource: a credential is required");
    // Before the mic and before the socket: a refused credential must not open
    // a device or a connection.
    const refusal = insecureCredentialRefusal({
      vendor: "GeminiLiveVoiceSource",
      isEphemeral: this.isEphemeral,
      allowInsecureApiKey: this.allowInsecureApiKey,
      ephemeralShape: "auth_tokens/…",
      mintHint: "POST https://generativelanguage.googleapis.com/v1beta/auth_tokens",
    });
    if (refusal) throw new Error(refusal);
    this.wantConnected = true;
    this.setState("initializing");
    try {
      // Gemini's rates are fixed and known up front, so the whole graph can
      // be up before the socket opens (ElevenLabs negotiates them instead).
      const mic = await PcmAudioGraph.requestMic();
      await this.graph.start({
        mic,
        inputRate: INPUT_RATE,
        outputRate: OUTPUT_RATE_FALLBACK,
        onChunk: (samples) => this.onCaptureChunk(samples),
      });
      await this.openSocket();
      this.connected = true;
      this.setState("listening");
      this.intervalId = setInterval(() => this.tick(), UPDATE_MS);
    } catch (err) {
      this.teardown();
      this.setState("idle");
      throw err;
    }
  }

  disconnect(): void {
    this.wantConnected = false;
    this.teardown();
    this.setState("idle");
  }

  /** An `auth_tokens/…` name is the ephemeral shape; anything else is a raw key. */
  private get isEphemeral(): boolean {
    return this.credential.startsWith("auth_tokens/");
  }

  private buildUrl(): string {
    return this.isEphemeral
      ? `${WS_BASE}${METHOD_TOKEN}?access_token=${encodeURIComponent(this.credential)}`
      : `${WS_BASE}${METHOD_KEY}?key=${encodeURIComponent(this.credential)}`;
  }

  private buildSetup(): Record<string, unknown> {
    const setup: Record<string, unknown> = {
      model: this.model.startsWith("models/") ? this.model : `models/${this.model}`,
      generationConfig: { responseModalities: ["AUDIO"] },
      // Gemini's own ASR of the user -- the vendor-side "the user is being
      // heard" signal this adapter uses instead of an energy heuristic.
      inputAudioTranscription: {},
      sessionResumption: this.resumptionHandle ? { handle: this.resumptionHandle } : {},
      contextWindowCompression: { slidingWindow: {} },
    };
    if (this.instructions) setup.systemInstruction = { parts: [{ text: this.instructions }] };
    return setup;
  }

  /** Opens the socket, sends `setup`, resolves on `setupComplete`. */
  private openSocket(): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      const ws = new WebSocket(this.buildUrl());
      // ArrayBuffer, not Blob, so every frame is decoded synchronously and
      // message order can never be reshuffled by an awaited Blob.text().
      ws.binaryType = "arraybuffer";
      this.ws = ws;
      let settled = false;
      const timer = setTimeout(() => {
        if (settled) return;
        settled = true;
        reject(new Error("Gemini Live setup did not complete within 15s"));
        ws.close();
      }, SETUP_TIMEOUT_MS);

      ws.onopen = () => ws.send(JSON.stringify({ setup: this.buildSetup() }));
      ws.onmessage = (e) => {
        const msg = parseServerMessage(e.data);
        if (!msg) return;
        if (msg.setupComplete !== undefined && !settled) {
          settled = true;
          clearTimeout(timer);
          this.streaming = true;
          resolve();
          return;
        }
        this.onServerMessage(msg);
      };
      ws.onerror = () => {
        if (settled) return;
        settled = true;
        clearTimeout(timer);
        reject(new Error("Gemini Live WebSocket failed to connect (check the credential and network)"));
      };
      ws.onclose = (ev) => {
        if (!settled) {
          settled = true;
          clearTimeout(timer);
          reject(new Error(`Gemini Live closed during setup (code ${ev.code}${ev.reason ? `: ${ev.reason}` : ""})`));
          return;
        }
        if (this.ws === ws) this.onSocketClosed(ev);
      };
    });
  }

  private onCaptureChunk(samples: Float32Array): void {
    if (!this.streaming || !this.ws || this.ws.readyState !== WebSocket.OPEN) return;
    this.ws.send(
      JSON.stringify({
        realtimeInput: {
          audio: { data: bytesToBase64(float32ToPcm16(samples)), mimeType: `audio/pcm;rate=${INPUT_RATE}` },
        },
      })
    );
  }

  private onServerMessage(msg: ServerMessage): void {
    const sc = msg.serverContent;
    if (sc) {
      if (sc.interrupted) {
        // The Live guide: "stop playing audio and clear queued playback".
        // An interrupt only if there was output to cut -- audible, or
        // already queued on the playback timeline.
        if (this.state === "speaking" || this.graph.playbackState() !== "drained") this.interruptCb?.();
        this.graph.clearPlayback(true);
        this.generationDone = true;
        this.setState("listening");
      }
      if ((sc.interimInputTranscription || sc.inputTranscription) && this.state !== "speaking") {
        this.setState("listening");
      }
      for (const part of sc.modelTurn?.parts ?? []) {
        const inline = part.inlineData;
        if (inline?.data && (inline.mimeType ?? "").startsWith("audio/pcm")) {
          this.graph.enqueue(pcm16ToFloat32(base64ToBytes(inline.data)), parsePcmRate(inline.mimeType, OUTPUT_RATE_FALLBACK));
          this.generationDone = false;
          // Received, not audible yet -- tick() promotes to `speaking` once
          // the playback clock actually reaches the burst.
          if (this.state !== "speaking") this.setState("thinking");
        }
      }
      if (sc.generationComplete || sc.turnComplete || sc.waitingForInput) this.generationDone = true;
    }
    const upd = msg.sessionResumptionUpdate;
    if (upd?.resumable && upd.newHandle) this.resumptionHandle = upd.newHandle;
    if (msg.goAway) {
      console.warn(`Gemini Live goAway (timeLeft ${msg.goAway.timeLeft ?? "?"}) -- reconnecting with the resumption handle`);
      void this.reconnect("goAway");
    }
    if (msg.error) console.error("Gemini Live error message:", msg.error);
  }

  private tick(): void {
    const metrics = this.graph.read();
    if (metrics) this.metricsCb?.(metrics);

    if (!this.connected) return; // reconnecting: stay `initializing`
    const playback = this.graph.playbackState();
    if (playback === "audible") {
      this.setState("speaking");
    } else if (playback === "drained") {
      // Drained. Still generating (buffer starved by the network) reads as
      // `thinking`, not `speaking` -- nothing is audible; done reads as
      // `listening`. Both only apply once we were actually in a turn.
      if (this.state === "speaking" || this.state === "thinking") {
        this.setState(this.generationDone ? "listening" : "thinking");
      }
    }
  }

  private onSocketClosed(ev: CloseEvent): void {
    if (!this.wantConnected) return; // user disconnect, already handled
    console.warn(`Gemini Live socket closed (code ${ev.code}${ev.reason ? `: ${ev.reason}` : ""})`);
    void this.reconnect("close");
  }

  private async reconnect(reason: string): Promise<void> {
    if (this.reconnecting || !this.wantConnected) return;
    this.reconnecting = true;
    this.streaming = false;
    this.connected = false;
    this.graph.clearPlayback(false);
    this.generationDone = true;
    this.setState("initializing");
    const old = this.ws;
    this.ws = null;
    if (old) {
      old.onopen = null;
      old.onmessage = null;
      old.onerror = null;
      old.onclose = null;
      try {
        old.close(1000);
      } catch {
        /* already closed */
      }
    }
    for (let attempt = 1; attempt <= RECONNECT_ATTEMPTS && this.wantConnected; attempt++) {
      try {
        await this.openSocket();
        this.connected = true;
        this.reconnecting = false;
        this.setState("listening");
        return;
      } catch (err) {
        console.warn(`Gemini Live reconnect ${attempt}/${RECONNECT_ATTEMPTS} after ${reason} failed:`, err);
        await new Promise((r) => setTimeout(r, RECONNECT_DELAY_MS));
      }
    }
    this.reconnecting = false;
    if (this.wantConnected) {
      console.error("Gemini Live: gave up reconnecting");
      this.disconnect();
    }
  }

  private setState(s: AgentState): void {
    if (this.state === s) return;
    this.state = s;
    this.stateCb?.(s);
  }

  private teardown(): void {
    this.connected = false;
    this.streaming = false;
    this.reconnecting = false;
    if (this.intervalId != null) clearInterval(this.intervalId);
    this.intervalId = null;
    if (this.ws) {
      const ws = this.ws;
      this.ws = null;
      ws.onopen = null;
      ws.onmessage = null;
      ws.onerror = null;
      ws.onclose = null;
      try {
        ws.close(1000);
      } catch {
        /* already closed */
      }
    }
    this.graph.stop();
    this.generationDone = true;
    this.resumptionHandle = null; // a fresh connect() starts a fresh session
  }
}

function parseServerMessage(data: unknown): ServerMessage | null {
  let text: string;
  if (typeof data === "string") text = data;
  else if (data instanceof ArrayBuffer) text = new TextDecoder().decode(data);
  else return null;
  try {
    return JSON.parse(text) as ServerMessage;
  } catch {
    return null;
  }
}

File: packages/android/gemini/src/main/kotlin/dev/sinua/gemini/GeminiLiveVoiceSource.kt
package dev.sinua.gemini

import android.util.Log
import dev.sinua.voice.AgentState
import dev.sinua.voice.AndroidPcmAudioDevice
import dev.sinua.voice.GeminiLiveSession
import dev.sinua.voice.InsecureCredential
import dev.sinua.voice.LiveSocket
import dev.sinua.voice.LiveSocketFactory
import dev.sinua.voice.LooperMainDispatcher
import dev.sinua.voice.MainDispatcher
import dev.sinua.voice.PcmAudioDevice
import dev.sinua.voice.PcmAudioGraph
import dev.sinua.voice.VoiceMetrics
import dev.sinua.voice.VoiceSource
import dev.sinua.websocket.OkHttpLiveSocketFactory

/**
 * `VoiceSource` for Gemini Live over its raw WebSocket -- the native mirror of
 * the Web `GeminiLiveVoiceSource` (packages/voice/src/GeminiLiveVoiceSource.ts)
 * and iOS `SinuaGeminiLive`. Protocol and state rules live in
 * `GeminiLiveSession`, the playback-timeline gate in `PcmAudioGraph` (both
 * dev.sinua.voice, JVM-tested); this class wires them to OkHttp and an audio
 * device. Callbacks arrive on the main thread.
 *
 * Mic PCM16 @ 16 kHz goes up; the model's PCM16 @ 24 kHz is scheduled on the
 * player; `speaking` and the metrics follow what has actually played. Session
 * resumption + reconnect on `goAway` or an unexpected close (3 attempts, 500 ms
 * apart), as on Web.
 *
 * `connect()` is synchronous on Android (`VoiceSource`): it opens the socket in
 * the background and returns. Only after `setupComplete` does it start the
 * audio (mic + player), so a bad or expired credential fails without ever
 * opening the mic. Any failure after `connect()` returned (setup, or starting
 * the audio, e.g. no RECORD_AUDIO) goes to `onError`, with the state already
 * back at `idle`.
 *
 * Credentials: an `auth_tokens/…` ephemeral token minted by your backend
 * (production shape), or -- dev only -- a raw API key. Both travel as request
 * headers, never in the URL; held in memory only, never logged.
 *
 * Not verified against the live API yet (docs/audio-pipeline.md).
 */
class GeminiLiveVoiceSource(
    credential: String,
    model: String = GeminiLiveSession.DEFAULT_MODEL,
    instructions: String? = null,
    /**
     * Opt in to a raw, long-lived API key. Without it `connect()` refuses one
     * before the socket or the mic ([InsecureCredential]). Local demos only.
     */
    private val allowInsecureApiKey: Boolean = false,
    /** Override the Google endpoint (a relay your backend runs, or tests). */
    endpoint: GeminiLiveSession.Endpoint? = null,
    device: PcmAudioDevice = AndroidPcmAudioDevice(),
    private val socketFactory: LiveSocketFactory = OkHttpLiveSocketFactory(),
    private val main: MainDispatcher = LooperMainDispatcher(),
) : VoiceSource {
    private val hasCredential = credential.isNotBlank()
    private val credentialIsEphemeral = InsecureCredential.isGeminiEphemeral(credential)
    private val endpoint = endpoint ?: GeminiLiveSession.endpoint(credential)
    private val session = GeminiLiveSession(model, instructions)
    private val graph = PcmAudioGraph(device)

    // Main-thread state.
    private var socket: LiveSocket? = null

    @Volatile private var micSocket: LiveSocket? = null // read on the capture thread
    private var setupPending: SetupWait? = null
    private var wantConnected = false
    private var reconnecting = false
    private var audioStarted = false
    private val pendingAudio = mutableListOf<Pair<FloatArray, Int>>()
    private var ticking = false
    private var metricsCb: ((VoiceMetrics) -> Unit)? = null
    private var errorCb: ((Throwable) -> Unit)? = null

    private class SetupWait(val onReady: () -> Unit, val onFail: (Throwable) -> Unit, val timeout: Runnable)

    private val ticker = object : Runnable {
        override fun run() {
            if (!ticking) return
            graph.read()?.let { metricsCb?.invoke(it) }
            if (!reconnecting) session.tick(graph.playbackState())
            main.postDelayed(this, (1000 / UPDATE_HZ).toLong())
        }
    }

    override fun onMetrics(cb: (VoiceMetrics) -> Unit) {
        metricsCb = cb
    }

    override fun onStateChange(cb: (AgentState) -> Unit) {
        session.onState = cb
    }

    override fun onInterrupt(cb: () -> Unit) {
        session.onInterrupt = cb
    }

    /** Setup / reconnect failures after `connect()` returned (the state is already back to idle). */
    fun onError(cb: (Throwable) -> Unit) {
        errorCb = cb
    }

    /** Call on the main thread. */
    override fun connect() {
        check(hasCredential) { "GeminiLiveVoiceSource: a credential is required" }
        // Before the socket, before the mic: a refused credential must not open
        // a device or a connection.
        InsecureCredential.check(
            vendor = "GeminiLiveVoiceSource",
            isEphemeral = credentialIsEphemeral,
            allowInsecureApiKey = allowInsecureApiKey,
            ephemeralShape = InsecureCredential.GEMINI_SHAPE,
            mintHint = InsecureCredential.GEMINI_MINT_HINT,
        )
        if (wantConnected) return
        wantConnected = true
        session.reset()
        session.connecting()
        // Authenticate first; the audio (and the mic) only after setupComplete.
        openSocket(
            onReady = { startAudio() },
            onFail = { err ->
                teardown()
                errorCb?.invoke(err)
            },
        )
    }

    private fun startAudio() {
        try {
            graph.start(GeminiLiveSession.INPUT_RATE, GeminiLiveSession.OUTPUT_RATE) { samples ->
                micSocket?.send(GeminiLiveSession.micMessage(samples))
            }
        } catch (e: Throwable) {
            teardown()
            errorCb?.invoke(e)
            return
        }
        audioStarted = true
        pendingAudio.forEach { (samples, rate) -> graph.enqueue(samples, rate) }
        pendingAudio.clear()
        micSocket = socket
        startTicker()
    }

    override fun disconnect() {
        teardown()
    }

    // --- socket (main thread) ---

    private fun openSocket(onReady: () -> Unit, onFail: (Throwable) -> Unit) {
        if (!wantConnected) return
        lateinit var opened: LiveSocket
        val timeout = Runnable {
            if (socket ===
                opened
            ) {
                failSetup(IllegalStateException("Gemini Live setup did not complete within 15s"))
            }
        }
        setupPending = SetupWait(onReady, onFail, timeout)
        opened = socketFactory.open(
            endpoint.url,
            endpoint.headers,
            emptyList(),
            onText = { text -> main.post { if (socket === opened) onText(text) } },
            onClose = { err -> main.post { if (socket === opened) onSocketClosed(err) } },
        )
        socket = opened
        opened.send(session.setupMessage())
        main.postDelayed(timeout, SETUP_TIMEOUT_MS)
    }

    private fun onText(text: String) {
        for (action in session.handle(text, graph.playbackState())) {
            when (action) {
                GeminiLiveSession.Action.SetupComplete -> {
                    if (audioStarted) micSocket = socket // a reconnect; the first connect sets it in startAudio
                    val wait = setupPending ?: continue
                    setupPending = null
                    main.remove(wait.timeout)
                    wait.onReady()
                }

                // Model audio between setupComplete and the audio starting is held, not dropped.
                is GeminiLiveSession.Action.Enqueue ->
                    if (audioStarted) {
                        graph.enqueue(action.samples, action.rate)
                    } else {
                        pendingAudio +=
                            action.samples to action.rate
                    }

                is GeminiLiveSession.Action.ClearPlayback -> {
                    pendingAudio.clear()
                    graph.clearPlayback(action.fade)
                }

                is GeminiLiveSession.Action.Reconnect -> reconnect(action.reason)

                is GeminiLiveSession.Action.ServerError -> log("Gemini Live error message: ${action.message}")
            }
        }
    }

    private fun onSocketClosed(err: Throwable?) {
        if (setupPending != null) {
            failSetup(err ?: IllegalStateException("Gemini Live closed during setup"))
            return
        }
        if (wantConnected) reconnect("close: ${err?.message ?: "closed"}")
    }

    private fun failSetup(err: Throwable) {
        val wait = setupPending ?: return
        setupPending = null
        main.remove(wait.timeout)
        closeSocket()
        wait.onFail(err)
    }

    private fun closeSocket() {
        micSocket = null
        val s = socket
        socket = null
        s?.close()
    }

    private fun reconnect(reason: String) {
        if (reconnecting || !wantConnected) return
        reconnecting = true
        closeSocket()
        graph.clearPlayback(false)
        session.connecting()
        attempt(1, reason)
    }

    private fun attempt(n: Int, reason: String) {
        if (!wantConnected) {
            reconnecting = false
            return
        }
        openSocket(
            onReady = { reconnecting = false },
            onFail = { err ->
                log("Gemini Live reconnect $n/$RECONNECT_ATTEMPTS after $reason failed: ${err.message}")
                if (n < RECONNECT_ATTEMPTS) {
                    main.postDelayed({ attempt(n + 1, reason) }, RECONNECT_DELAY_MS)
                } else {
                    reconnecting = false
                    log("Gemini Live: gave up reconnecting")
                    teardown()
                    errorCb?.invoke(err)
                }
            },
        )
    }

    // --- tick / teardown ---

    private fun startTicker() {
        if (ticking || !wantConnected) return
        ticking = true
        main.post(ticker)
    }

    private fun teardown() {
        wantConnected = false
        reconnecting = false
        ticking = false
        main.remove(ticker)
        setupPending?.let { main.remove(it.timeout) }
        setupPending = null
        closeSocket()
        audioStarted = false
        pendingAudio.clear()
        graph.stop()
        session.reset() // idle; a fresh connect() starts a fresh session
    }

    private fun log(msg: String) {
        try {
            Log.w(TAG, msg)
        } catch (_: RuntimeException) {
            // android.util.Log is a stub under plain JVM tests.
        }
    }

    companion object {
        const val UPDATE_HZ = 30.0
        private const val TAG = "GeminiLive"
        const val SETUP_TIMEOUT_MS = 15_000L
        const val RECONNECT_ATTEMPTS = 3
        const val RECONNECT_DELAY_MS = 500L
    }
}
